akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

```diff
@@ -693,3 +699,470 @@ private[akka] class BroadcastHub[T](bufferSize: Int) extends GraphStageWithMater
+
+        case TryPull ⇒
+          tryPull()
+
+        case RegistrationPending ⇒
+          state.getAndSet(noRegistrationsState).asInstanceOf[Open].registrations foreach { consumer ⇒
+            consumers :+= consumer
+            consumers = consumers.sortBy(_.id)
+            consumerIds :+= consumer.id
+            consumerIds.sorted
+            queue.init(consumer.id)
+            if (consumers.size >= startAfterNbrOfConsumers) {
+              initialized = true
+            }
+
+            consumer.callback.invoke(Initialize)
+
+            if (initialized && pending.nonEmpty) {
+              pending.foreach(publish)
+              pending = Vector.empty[T]
+            }
+
+            tryPull()
+          }
+
+        case UnRegister(id) ⇒
+          consumers = consumers.filterNot(_.id == id)
+          consumerIds = consumerIds.filterNot(_ == id)
+          queue.remove(id)
+          if (consumers.isEmpty) {
+            if (isClosed(in)) completeStage()
+          } else
+            tryPull()
+      }
+    }
+
+    override def onUpstreamFailure(ex: Throwable): Unit = {
+      val failMessage = HubCompleted(Some(ex))
+
+      // Notify pending consumers and set tombstone
+      state.getAndSet(Closed(Some(ex))).asInstanceOf[Open].registrations foreach { consumer ⇒
+        consumer.callback.invoke(failMessage)
+      }
+
+      // Notify registered consumers
+      consumers.foreach { consumer ⇒
+        consumer.callback.invoke(failMessage)
+      }
+      failStage(ex)
+    }
+
+    override def postStop(): Unit = {
+      // Notify pending consumers and set tombstone
+
+      @tailrec def tryClose(): Unit = state.get() match {
+        case Closed(_) ⇒ // Already closed, ignore
+        case open: Open ⇒
+          if (state.compareAndSet(open, Closed(None))) {
+            val completedMessage = HubCompleted(None)
+            open.registrations foreach { consumer ⇒
+              consumer.callback.invoke(completedMessage)
+            }
+          } else tryClose()
+      }
+
+      tryClose()
+    }
+
+    // Consumer API
+    def poll(id: Long, hubCallback: AsyncCallback[HubEvent]): AnyRef = {
+      // try pull via async callback when half full
+      // this is racy with other threads doing poll but doesn't matter
+      if (queue.size == DemandThreshold)
+        hubCallback.invoke(TryPull)
+
+      queue.poll(id)
+    }
+
+    setHandler(in, this)
+  }
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Source[T, NotUsed]) = {
+    val idCounter = new AtomicLong
+
+    val logic = new PartitionSinkLogic(shape)
+
+    val source = new GraphStage[SourceShape[T]] {
+      val out: Outlet[T] = Outlet("PartitionHub.out")
+      override val shape: SourceShape[T] = SourceShape(out)
+
+      override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with OutHandler {
+        private val id = idCounter.getAndIncrement()
```

Comment:
Materialized value of the identifier doesn't work since it's the function that needs it, so it's too late (unless you set it into some mutable variable, but that would be weird).

The identifier is a counter that is incremented in the same order as the consumers are attached/materialized. Perhaps that is enough?